Create a custom thrift protocol to ensure client/server compatibility

DomGarguilo opened the pull request:
Addresses #5390

DomGarguilo committed:
Create a custom thrift protocol to ensure client/server compatibility

DomGarguilo commented:
Created this in draft since, for now, it only addresses the task outlined in the ticket:

> create a custom protocol with the versioning information stored in a header and checks on the server side to validate it

I'm not sure if any of the other proposed tasks should be added here or as follow-on PRs.

DomGarguilo committed:
Fix checkstyle violation

keith-turner commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,156 @@
+  public static class AccumuloProtocol extends TCompactProtocol {
+
+    private static final int MAGIC_NUMBER = 0x41434355; // "ACCU" in ASCII
+    private static final byte PROTOCOL_VERSION = 1;
```
Should probably add a comment to AccumuloDataVersion.CURRENT_VERSION that references this and mentions that when changing CURRENT_VERSION may also need to change this.

ctubbsii replied:
I was wondering about that... whether the protocol version should be independent, so we can version the serialization format of the protocol header.

We will probably want to include the `AccumuloDataVersion.CURRENT_VERSION` in the protocol header directly, though. We could also include the actual Accumulo release version in here, too (major.minor.patch; e.g. 4.0.0). I think to verify, we'd probably want to ensure the same major and same minor version, if we want to be even more restrictive than the data version. We really shouldn't have Accumulo servers or clients/servers talking across major.minor versions (different patch releases are okay, and is important for rolling upgrades).

DomGarguilo replied:
I left it separate for now just in case we wanted to have multiple protocol versions able to talk to each other but am not sure about that. 

The data version and protocol version definitely seem related but I'm not sure how closely tied we want to have those versions.

ctubbsii replied:
I think the protocol version can be independently numbered, not connected to the data version. I think it's sufficient to check the Accumulo version (Constants.VERSION) major.minor. We don't need to use the data version in here at all. The protocol version is the version of the RPC protocol... the data version is the version of the data persisted in ZK and HDFS... and the Accumulo version is the overall version of the software. Only the RPC protocol version and the Accumulo version are relevant here.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,156 @@
+        try {
+          this.writeHeader();
+        } catch (TException e) {
+          if (scope != null) {
```
span has a recordException message, but not sure if that should be used here.  I have seen it but have not looked into it in detail.  Might be worth looking into.

DomGarguilo replied:
I took a look at the [Exceptions](https://opentelemetry.io/docs/specs/otel/trace/exceptions/#recording-an-exception) and [Recording errors](https://opentelemetry.io/docs/specs/semconv/general/recording-errors/) pages on the open telemetry docs and pushed commit 5dfab38 with what seems like the correct approach.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,156 @@
+        }
+      }
+
+      super.writeMessageBegin(message);
```
Seems like this should be inside the try block above.  Thinking if this message throws exception that writeMessageEnd will never be called.  If it should be in the try block then maybe the existing code in 2.1 has a bug.

DomGarguilo replied:
I'm not sure I understand. Putting `super.writeMessageBegin(message);` in the try-catch would close the span and scope if an exception is thrown from it but I'm not sure that it would make `writeMessageEnd()` in either case, right? Maybe im missing something here.

keith-turner replied:
If `super.writeMessageBegin(message)` throws an exception, then it we can probably assume that `writeMessageEnd()` will never be called. So if this happens will nothing close the span and scope?

DomGarguilo replied:
Oh yea, that sounds correct. I misunderstood what you were saying. I'll make sure span and scope get closed in that case.

DomGarguilo replied:
Should be fixed as of 1ea49cc

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,156 @@
+    public void writeMessageEnd() throws TException {
+      super.writeMessageEnd();
+
+      if (this.isClient && scope != null) {
+        scope.close();
+        span.end();
+      }
```
I don't think we care if it's a client. If a scope was opened, we'd want to close it.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,156 @@
+
+    @Override
+    public void writeMessageBegin(TMessage message) throws TException {
+      if (this.isClient) {
```
The code this replaces used the same protocol for both clients and servers. I'm not 100% sure when writeMessageBegin gets called, but I think it's probably fine to do it on every message, and not discriminate based on client and server.

DomGarguilo replied:
Yea I am kinda unsure about this point (not sure whats correct/better). 

If I understand things correctly here, the thrift messages flow in both directions between "client" and "server" over the same connection. So as long as we validate compatibility when the client sends a message to the server, than we can avoid doing the extra work of adding (and therefore validating) the headers when the server responds to the client.

I think things would work just fine if we did not distinguish between client and server in this context, we would just be doing more work than necessary which might be unoptimal (which seems like it matters in this context).

I'm also thinking though there might be the benefit of always writing and validating these header just so we are certain that there are no edge cases or scenarios that I'm not thinking about where a server responds to a client over a connection that has not been validated yet. I don't know if this would ever happen but it seems like a trade off between absolute certainty that we have valid comms both ways and optimizing the amount of work done (again might be worth the optimization in this case, not sure).

Definitely would appreciate feedback on this line of thinking and corrections regarding any incorrect assumptions ive made above.

keith-turner replied:
> If I understand things correctly here, the thrift messages flow in both directions between "client" and "server" over the same connection. So as long as we validate compatibility when the client sends a message to the server, than we can avoid doing the extra work of adding (and therefore validating) the headers when the server responds to the client.

That sounds good to me.  It seems like the changes in the code to support this are minimal, do you know if you have caught everything in the code for covering this functionality?

DomGarguilo replied:
> do you know if you have caught everything in the code for covering this functionality?

As far as I can tell, yes. It seems like our code either uses `ThriftUtil` or `TServerUtils` to construct these objects and I've made sure those places use this new code.

ctubbsii replied:
> I think things would work just fine if we did not distinguish between client and server in this context, we would just be doing more work than necessary which might be unoptimal (which seems like it matters in this context).

Your explanation makes sense to me. I was just comparing to what was there before, and what was there seemed like it didn't discriminate when setting up the trace spans.

ctubbsii reviewed (commented):
I'm not sure if we actually need to distinguish between client and server for the spans.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
Address feedback

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
add exception recording to the span

DomGarguilo committed:
remove TInfo and use headers instead

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
make header map final

DomGarguilo committed:
remove missed TInfo use

DomGarguilo commented:
Marking this as ready for review. I tried to address the following two suggestions from the ticket:

> 1. Primarily, create a custom protocol with the versioning information stored in a header and checks on the server side to validate it, then
> 2. Secondarily, simplify the thrift APIs by adding tracing info to the protocol header

There are a lot of changed files in this PR since I had to update the thrift code. There are also a lot of places where the only change is the removal of the `TInfo` object as the first param in a method call which was a trivial change. Hopefully that helps skip/skim over portions of the diff while reviewing.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

keith-turner commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,236 @@
+        W3CTraceContextPropagator.getInstance().inject(Context.current(), traceHeaders,
+            (headers, key, value) -> headers.put(key, value));
+
+        super.writeI16((short) traceHeaders.size());
```
Should probably ensure traceHeaders.size() is less than Short.MAX_VALUE for safety, otherwise it could go negative.  However since this class extends TCompactProtocol could call `writeI32` instead which in TCompactProtocol will call `writeVarint32` which will make the integer as short as possible.   The impl of `writeI16` also calls `writeVarint32` so it does not seem there is much advantage to calling `writeI16` in this case.

DomGarguilo replied:
Fixed in 59a55ae. Seems like using `writeI32` makes the most sense here.

DomGarguilo committed:
Add unit test

DomGarguilo committed:
Use int instead of short

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

kevinrr888 commented on core/src/main/java/org/apache/accumulo/core/rpc/ThriftUtil.java:
```diff
@@ -63,12 +66,16 @@ public class ThriftUtil {
   /**
-   * An instance of {@link TraceProtocolFactory}
+   * An instance of {@link AccumuloProtocolFactory}
    *
    * @return The default Thrift TProtocolFactory for RPC
    */
-  public static TProtocolFactory protocolFactory() {
-    return protocolFactory;
+  public static TProtocolFactory clientProtocolFactory() {
+    return clientAccumuloProtocolFactory;
+  }
+
+  public static TProtocolFactory serverProtocolFactory() {
+    return serverAccumuloProtocolFactory;
```
This @ return may need to be updated. Should also probably have a similar javadoc for `serverProtocolFactory()`

DomGarguilo commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,236 @@
+            + Integer.toHexString(MAGIC_NUMBER) + ", got: 0x" + Integer.toHexString(magic));
+      }
+
+      final byte version = super.readByte();
+      if (!isCompatibleVersion(version)) {
+        throw new TException("Incompatible protocol version. Client version: " + version
+            + ", Server version: " + PROTOCOL_VERSION);
+      }
```
In this case its always client -> server. In this context "client" just means the sending entity and "server" means the entity that receives the message. I do think this can be confusing though and it might be better to just remove "client" and "server" from the message since thats not really helpful. I'll revise the exception message to more closely align with your suggestion.

kevinrr888 replied:
Ah, I see that makes sense then. The updated message also seems good.

DomGarguilo committed:
Address feedback

DomGarguilo commented:
@kevinrr888 your suggestions should all be addressed as of 29c1234

in that commit I...
* refactored the protocol version compatibility check method to throw an exception directly instead of return a boolean. I also updated the wording on that message to make things more clear
* addressed the IDE warnings regarding nullness in various places
* added javadoc to the static client and server factory methods in ThriftUtil to make it clearer which should be used

DomGarguilo committed:
Use TriftUtil helper method for exception tracing

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,241 @@
+            span.recordException(e);
+            span.setAttribute("error.type", e.getClass().getCanonicalName());
+            span.setStatus(StatusCode.ERROR, e.getMessage());
+          }
```
Nice find! I originally kept it separate since it looked a bit different but on closer inspection those two spots were doing the same thing so I just used the mentioned helper method. Addressed in 473426d

ctubbsii commented:
Will try to review this week.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,238 @@
+      super.writeI32(MAGIC_NUMBER);
+      super.writeByte(PROTOCOL_VERSION);
+
+      if (this.isClient && span != null && span.getSpanContext().isValid()) {
```
This method is only called when isClient is true, so the check seems misleading and unnecessary.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,238 @@
+        scope = span.makeCurrent();
+
+        try {
+          this.writeHeader();
```
why use `this.` here?  also used in some other places in this class

DomGarguilo replied:
Its just preference. I think it reads more like a sentence when reading the code and makes it more apparent that the writeHeader method is operating on the object itself. I am fine with removing it if you think its better off without it.

keith-turner replied:
It does not bother me.  I was asking because I was not sure if it was required for some reason that I did not understand.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,238 @@
+        super.writeBool(HEADER_HAS_TRACE);
+        traceHeaders.clear();
+
+        W3CTraceContextPropagator.getInstance().inject(Context.current(), traceHeaders,
```
Assuming this code was copied from TraceUtil?  Was there any significant change you needed to make to it to use it here?

DomGarguilo replied:
Yea it was pretty much just copied from TraceUtil.
```java
  public static TInfo traceInfo() {
    TInfo tinfo = new TInfo();
    W3CTraceContextPropagator.getInstance().inject(Context.current(), tinfo, TInfo::putToHeaders);
    return tinfo;
  }
```
The main difference is the use of the TInfo object instead of just a regular map in the new code.

keith-turner replied:
> I ended up refactoring out these inject and extract code blocks into methods in TraceUtil in https://github.com/apache/accumulo/commit/2c4eff6c50e9e1e07351400967ca6ab6ea2e88c8. Thats where they were before and I think its fine to keep them like that.

That is nice in that it makes the code in AccumuloProtocol more focused on its main task.

keith-turner commented on core/src/test/java/org/apache/accumulo/core/rpc/AccumuloProtocolTest.java:
```diff
@@ -0,0 +1,120 @@
+      var serverProtocol = (AccumuloProtocolFactory.AccumuloProtocol) AccumuloProtocolFactory
+          .serverFactory().getProtocol(transport);
+
+      assertDoesNotThrow(serverProtocol::validateHeader);
```
Would be good to verify all the data was read from `transport`.  Experimented w/ this and found the following seems to work.  There are some other methods on `transport` that sound promising based on their name, but do not actually verify the data was read.

```suggestion
      assertDoesNotThrow(serverProtocol::validateHeader);
      
        // All data should have been read from the transport following should read zero bytes.
      assertEquals(0, transport.read(new byte[10],0,10));    
```

DomGarguilo replied:
Added in 835cbf5

keith-turner reviewed (approved):
These changes look good. I kinda get lost when I look at this w/ all the TInfo changes.  Are there any classes in particular that you want reviewed besides AccumuloProtocol?

Not sure if we have any ITs that cover tracing, do you know if we do?  If not did you try manually testing tracing w/ these changes?

DomGarguilo commented:
> I kinda get lost when I look at this w/ all the TInfo changes. Are there any classes in particular that you want reviewed besides AccumuloProtocol?

Yea there are a lot of changed files from getting rid of the TInfo object. The other changes here are mostly limited to AccumuloProtocolFactory, ThriftUtil and TServerUtils and even then there is not a whole lot that has changed outside of AccumuloProtocolFactory.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
Refactor to remove redundant isClient check and avoid nesting

DomGarguilo committed:
Refactor trace extraction/injection code back into TraceUtil

DomGarguilo committed:
add to test case to ensure everything was read from protocol

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

keith-turner commented:
@DomGarguilo did not realize merging the ClientTabletCache changes would conflict w/ this.   Makes sense though w/ all the TInfo changes.  Would probably be good to get this merged in, the testing that was done sounds really good.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,220 @@
+     * @throws TException if the given protocol version is incompatible with the current version
+     */
+    private void validateProtocolVersion(byte protocolVersion) throws TException {
+      if (protocolVersion != PROTOCOL_VERSION) {
+        throw new TException("Incompatible protocol version. Version seen: " + protocolVersion
+            + ", expected version: " + PROTOCOL_VERSION);
```
This requires us to keep the protocol version updated whenever we change the thrift RPC... however, I think we need to proactively prevent communication with any potentially incompatible version of Accumulo.

We generally try to support rolling upgrades for bugfix releases, but for all other upgrades, we expect the whole cluster to be shut down and restarted with the new version. With that in mind, I think we should serialize Constants.VERSION to the protocol, and we should validate by matching on the first two elements of the Accumulo version to ensure it matches the server's own first two elements of its Constants.VERSION.

We can still have the protocol version in case we change how things are serialized in the protocol (such as adding another header field prior to the serialized trace context or we want to change the boolean to an int or something)... but we won't need to bump it very often unless we change the protocol itself. We shouldn't need to bump it just because of an Accumulo version release, but we still need to check the Accumulo version.

DomGarguilo replied:
Added this in 713226c. Right now it splits on the last `.` character and checks that the string before that is the same on both ends. It could be helpful to add some standard way to get the parts of the version number if we find that this method does not work like we expect.

ctubbsii replied:
The build helper maven plugin can split up the components in a standard way. I can make that change to the build after this is done. It should be trivial.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/trace/TraceUtil.java:
```diff
@@ -131,48 +176,6 @@ public static void setException(Span span, Throwable e, boolean rethrown) {
-   * is used to send the current trace information to a remote process
-   */
-  public static TInfo traceInfo() {
-    TInfo tinfo = new TInfo();
-    W3CTraceContextPropagator.getInstance().inject(Context.current(), tinfo, TInfo::putToHeaders);
-    return tinfo;
-  }
-
-  /**
-   * Returns a newly created Context from the TInfo object sent by a remote process. The Context can
-   * then be used in this process to continue the tracing. The Context is used like:
-   *
-   * <pre>
-   * Context remoteCtx = getContext(tinfo);
-   * Span span = tracer.spanBuilder(name).setParent(remoteCtx).startSpan()
-   * </pre>
-   *
-   * @param tinfo tracing information serialized over Thrift
-   */
-  private static Context getContext(TInfo tinfo) {
-    return W3CTraceContextPropagator.getInstance().extract(Context.current(), tinfo,
-        new TextMapGetter<TInfo>() {
-          @Override
-          public Iterable<String> keys(TInfo carrier) {
-            if (carrier.getHeaders() == null) {
-              return null;
-            }
-            return carrier.getHeaders().keySet();
-          }
-
-          @Override
-          public String get(TInfo carrier, String key) {
-            if (carrier.getHeaders() == null) {
-              return null;
-            }
-            return carrier.getHeaders().get(key);
-          }
-        });
-  }
-
```
These two old methods were basically serialize/deserialize. We can replace them with new serialize/deserialize methods, even though the type is not TInfo, but some other serialized form (String).

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/trace/TraceUtil.java:
```diff
@@ -191,22 +194,13 @@ public static <T> Callable<T> unwrap(Callable<T> c) {
-      }
-      Span span = startServerRpcSpan(instance.getClass(), method.getName(), (TInfo) args[0]);
-      try (Scope scope = span.makeCurrent()) {
+      Span span = Span.current(); // should be set by protocol
+      try {
```
So, I think that what the old code was doing when it called `startServerRpcSpan` was that it was always creating a new "inner span" inside the outer span that that the user might have created in the client. That was useful because it helped us distinguish the a child or inner span that was just associated with the RPC call, rather than be grouped with all the other things the client code might have been doing around the RPC call.

This code, however, seems to just grab the current span the user created, and attach an exception to it if an RPC exception occurred. I think we should do the other thing. If there's an exception, it should be attached to the inner/child span for the RPC, not the outer span the user created.

DomGarguilo replied:
Okay I think I fixed this in 8d742ea. Please take a look.

ctubbsii replied:
I think I may have misunderstood what you were doing before. I think you might have had it correct before. I didn't realize you were creating the server span in the protocol, prior to this invocation handler being called. The expectation, I think, was that the invocation handler just needed to get a reference to the span started by the protocol, for the purposes of attaching any exceptions to it. I think I will need to look at this more closely, but I am now thinking I misled you by my earlier comment, and you were doing it correctly before.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
Rename to writeCLientHeader(), set scope & span null, remove fixed boolean const

Co-authored-by: Christopher Tubbs <ctubbsii@apache.org>

DomGarguilo committed:
Move serialize and deserialize to TraceUtil

DomGarguilo committed:
Rename trace from handleMessage to handleRpcMessage

Co-authored-by: Christopher Tubbs <ctubbsii@apache.org>

DomGarguilo commented on core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java:
```diff
@@ -0,0 +1,220 @@
+      super.writeBool(HEADER_HAS_TRACE);
+      traceHeaders.clear();
+
+      TraceUtil.injectTraceContext(traceHeaders);
```
Addressed in 5580e8d

DomGarguilo committed:
Validate against accumulo version

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
Improve error message and tests

DomGarguilo committed:
Fix unapproved char

DomGarguilo commented:
> One additional thought I had was that clients should send the instanceId, and the server should verify it, so that we don't get connections from clients for a different instance.

I was trying to work this in but am not sure the best place to pull in the instanceId from. I tried to pass it in as a parameter but had to reach pretty high up the chain which changed a lot of code. Was hoping you had a better idea for where I can add this in from.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
Create inner child span for each RPC

ctubbsii commented on core/src/test/java/org/apache/accumulo/core/rpc/AccumuloProtocolTest.java:
```diff
@@ -103,6 +105,59 @@ public void testIncompatibleVersion() throws TException {
+          (AccumuloProtocolFactory.AccumuloProtocol) AccumuloProtocolFactory.serverFactory()
+              .getProtocol(transport);
+
+      assertDoesNotThrow(serverProtocol::validateHeader,
+          "Expected compatible version to pass validation");
```
I think assertDoesNotThrow can be useful when we want to ignore a return value that normally would be thrown and expected to be used by some static analysis tool (like spotbugs). But in most code, we can just skip it altogether, as in:

```suggestion
      serverProtocol.validateHeader();
```

ctubbsii commented:
> > One additional thought I had was that clients should send the instanceId, and the server should verify it, so that we don't get connections from clients for a different instance.
> 
> I was trying to work this in but am not sure the best place to pull in the instanceId from. I tried to pass it in as a parameter but had to reach pretty high up the chain which changed a lot of code. Was hoping you had a better idea for where I can add this in from.

The ServerContext and/or ClientContext should be able to provide the InstanceId, and the protocol factory could be constructed with it. I'm not sure how much code that touches. I can try to take a look if you hit a wall here.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/trace/TraceUtil.java:
```diff
@@ -191,22 +250,18 @@ public static <T> Callable<T> unwrap(Callable<T> c) {
         throw t;
       } finally {
-        span.end();
+        if (childSpan != null) {
+          childSpan.end();
+        }
```
I don't think this can be null, unless the line 255 threw an NPE. The previous code didn't check for this and did not have an error. I think that perhaps this isn't possible.

DomGarguilo committed:
Update core/src/test/java/org/apache/accumulo/core/rpc/AccumuloProtocolTest.java

Co-authored-by: Christopher Tubbs <ctubbsii@apache.org>

DomGarguilo committed:
Remove unneeded null check

Co-authored-by: Christopher Tubbs <ctubbsii@apache.org

DomGarguilo committed:
Add instance ID validation to header

DomGarguilo committed:
use getInstanceId() method

DomGarguilo committed:
fix checkstyle

DomGarguilo commented:
> > > One additional thought I had was that clients should send the instanceId, and the server should verify it, so that we don't get connections from clients for a different instance.
> > 
> > 
> > I was trying to work this in but am not sure the best place to pull in the instanceId from. I tried to pass it in as a parameter but had to reach pretty high up the chain which changed a lot of code. Was hoping you had a better idea for where I can add this in from.
> 
> The ServerContext and/or ClientContext should be able to provide the InstanceId, and the protocol factory could be constructed with it. I'm not sure how much code that touches. I can try to take a look if you hit a wall here.

Okay I added the instance ID validation in 769d38f. Had to reach pretty far up the chain to get the instanceId so there were quite a few files changed but all of those are just parameter changes.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

ctubbsii commented:
Can you provide an example of the opentelemetry tracing output (screenshot or log sequences or something)? I'd be curious what they look like when a client performs a traced operation over the new RPC protocol.

DomGarguilo commented:
> Can you provide an example of the opentelemetry tracing output (screenshot or log sequences or something)? I'd be curious what they look like when a client performs a traced operation over the new RPC protocol.

I'm not too sure how much this shows but here is a screenshot from the "graph view" in Jaeger capturing traces from a scan from the shell
![image](https://github.com/user-attachments/assets/688dfe12-67f0-4147-8c2b-4a1d84b6641f)

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo commented:
> @DomGarguilo and @ctubbsii curious what the status of this PR is. Would be nice to get this merged in if possible.

Ready to be merged unless anyone else has blockers here.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo commented:
I am actually not sure that traces are being propagated correctly. I'm looking into things but I don't think this should be merged yet.

DomGarguilo committed:
Add override for readMessageEnd, closing span

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

DomGarguilo commented:
> I am actually not sure that traces are being propagated correctly. I'm looking into things but I don't think this should be merged yet.

After debugging. The issue is not fixed via a683811 and I can see spans from client -> server in Jaeger. The issue was that we were not closing the scope or span on `AccumuloProtocol.readMessageEnd()` so the traces were never closed on the server side, and thus those server spans never finished (and couldn’t be linked back to the client spans in Jaeger).

In the screenshot below, in a shell I ran `trace on` then `scan` (on an existing table with an entry) then `trace off`.
![image](https://github.com/user-attachments/assets/a814eecc-17f1-4edb-b475-29fcd3c85d49)

This demonstrates that cross-component tracing is working.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol

ctubbsii reviewed (commented):
Looking at the code again today, after that fix for the readMessageEnd...

I think the stuff with the version checks, and all that is all good, and we should definitely merge that stuff in soon. And, I really wanted the changes to the tracing stuff also, because I was really hoping it would allow us to simplify a lot of the Thrift code.

However, I'm starting to wonder if pushing the tracing stuff into this is actually feasible. When I looked at it again today, I think I'm starting to realize that readMessageEnd might finish before the actual action that the RPC was requesting is executed. If that's the case, then the span is already closed, and any subsequent spans on the server side are going to be disconnected. That seems to match what I'm seeing in the second screenshot, where we only see the handleRpcMessage span, but no child spans for it on the server.

The current code creates a span that wraps the entire execution of the RPC call, but I think in the protocol, it's only wrapping the message parsing, not the execution. I am not too familiar with Thrift internals, but I suspect a subsequent writeMessage is done on the server side to send a message back to the client with the results of the execution. If so, we really want to close the handleRpcMessage span at that time, and not before.

It may still be possible to salvage these tracing changes, and close the span at the end of the RPC execution inside the protocol factory. However, if that's not possible, we may have to abandon the thrift tracing changes, and continue doing that the way it is currently done, so it wraps the whole RPC operation, and only keep the portions that do the version/compatibility checks.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into accumuloProtocol
